spinoso-string/src/lib.rs

```diff
@@ -1649,18 +1651,19 @@ impl String {
     #[must_use]
     pub fn index<T: AsRef<[u8]>>(&self, needle: T, offset: Option<usize>) -> Option<usize> {
         fn inner(buf: &[u8], needle: &[u8], offset: Option<usize>) -> Option<usize> {
-            if let Some(offset) = offset {
-                let buf = buf.get(offset..)?;
-                let index = buf.find(needle)?;
-                // This addition is guaranteed not to overflow because the result is
-                // a valid index of the underlying `Vec`.
-                //
-                // `self.buf.len() < isize::MAX` because `self.buf` is a `Vec` and
-                // `Vec` documents `isize::MAX` as its maximum allocation size.
-                Some(index + offset)
-            } else {
-                buf.find(needle)
+            let offset = offset.unwrap_or(0);
+            let buf = std::str::from_utf8(buf).ok()?;
+            let needle = std::str::from_utf8(needle).ok()?;
```

Comment:
one thing I noticed here is that this requires strings to be valid UTF-8 which may not be true for strings of _any_ encoding.

I think this method is more appropriately delegated down through `EncodedString` to the individual string variants. Take a peek at `String::get_char_slice` to see how to implement something like that.

Reply:
Hi @lopopolo I've refactored most encoding-aware logic down to the encoded variants, so it's ready for review :)

The `enc::ascii::(r)index` and `enc::binary::(r)index` turn out to be quite duplicated, but it seems to be common across many `(Ascii|Binary)String` methods. Just want to confirm if there is anything I should do in this PR to DRY them?

Reply:
yup that's right, no DRYing up needed. that's the way I was expecting it to look.